Inspector remove relation

anthonydillon opened the pull request.

jujugui commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5148/
Test FAILed.

hatched commented:
QA Issues:
1. When selecting a relation to view its details it will navigate to those details as expected but it will also check the input just before navigating. It's a pretty trivial issue but it doesn't exist with the units lists.
2. With a deployed relation, Once you have removed it via the inspector there is no indication that it has been removed. I'm able to repeatedly remove the relation which creates multiple ECS entries which you can see in the commit summary

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5149/
Test PASSed.

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5150/
Test PASSed.

anthonydillon commented:
## QA
- Add wordpress and mariadb applications to the canvas
- Create a relation between them
- Click on one to the service
- Enter the Relations tab in the inpector
- Click on the relation name and check you see the relations details
- Go back to the relation list
- Check the Remove button is disabled
- Check the checkbox by the relation
- The Remove button should become active
- Click it and see the relation is removed and you are shown a message that there are no relations
- Add apache2 to the canvas
- Add relations between apache2 and mariadb to wordpress
- Deploy the changes
- Go into the relations for wordpress
- Check Select all and see that all relations are selected
- Click remove and see that all the selected relations are removed (The relation goes into uncommitted state on the canvas)
- Commit the changes
- Rebuild the two relations for wordpress
- Enter the relations list for wordpress
- Select a single relation
- Click Remove and see that only that one relation is removed

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5151/
Test PASSed.

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations/inspector-relations.js:
```diff
@@ -20,43 +20,177 @@ YUI.add('inspector-relations', function() {
+      Remove the selected relations
+
+      @method _handleRemoveRelation
+      @param {String} action The action to apply to the units.
```
Doesn't look like this is required anymore.

huwshimi commented:
@anthonydillon one QA issue:
- add mariadb and wordpress
- relate the services
- commit
- click on a service and click on relations
- check the relation and click remove
- you will see the message "No active relations for this application." but also the disabled Remove button

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations-item/_inspector-relations-item.scss:
```diff
@@ -1,34 +1,50 @@
-        svg {
-            vertical-align: text-bottom;
+    label {
+        padding: 11px 20px 11px 20px;
```
Could just be `padding: 11px 20px;` if you wanted.

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations-item/inspector-relations-item.js:
```diff
@@ -19,6 +19,20 @@ YUI.add('inspector-relations-item', function() {
 
+
+    /**
+      Get the current state of the inspector.
```
"Generate the initial state for the component."?

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations-item/inspector-relations-item.js:
```diff
@@ -38,19 +52,77 @@ YUI.add('inspector-relations-item', function() {
+        // When whenChanged is set by the list parent and is used to (de)select
+        // all checkboxes. It is called in the setState callback so that the
+        // updated state is available if we inspect it from whenChanged.
+        if (whenChanged) {
```
It looks like whenChanged is always defined so you might not need this check.

huwshimi replied:
Might be worth adding some PropTypes to this component and setting whenChanged to isRequired.

anthonydillon replied:
Yea agree. Removed the check and added a PropType.

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations-item/inspector-relations-item.js:
```diff
@@ -38,19 +52,77 @@ YUI.add('inspector-relations-item', function() {
+      @param {String} id The id of the checkbox.
+      @returns {String} The id of the element or a blank string.
+    */
+    _generateId: function(id) {
```
I'm actually not sure if we need to do this anymore, is this ever used as a navigation element?

huwshimi replied:
Oh, I see the select all doesn't have a relation, but we should probably still return an id for the `htmlFor`

anthonydillon replied:
Thing is we dont want the `htmlFor` to match the input because then it will tick the checkbox as it navigates to the details view. This is the same in the units view in the inspector.

huwshimi replied:
Ah yes, makes sense!

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations/inspector-relations.js:
```diff
@@ -20,43 +20,177 @@ YUI.add('inspector-relations', function() {
+      component.
+
+      @method _selectAllRelations
+      @param {String} group The key for a group of checkboxes.
```
You can remove this @param

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations/inspector-relations.js:
```diff
@@ -20,43 +20,177 @@ YUI.add('inspector-relations', function() {
+
+      @method _selectAllRelations
+      @param {String} group The key for a group of checkboxes.
+      @param {Boolean} checked Whether the "select all" child component is
```
"Whether to check or uncheck the relations"? Something along those lines...

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations/inspector-relations.js:
```diff
@@ -20,43 +20,177 @@ YUI.add('inspector-relations', function() {
+          });
+        }
+      });
+      if (checked === undefined) {
```
This doesn't look like it's doing anything...

anthonydillon replied:
Opps left over logic from removing the grouping functionality. Removed.

huwshimi commented on jujugui/static/gui/src/app/components/inspector-relations/inspector-relations.js:
```diff
@@ -20,43 +20,177 @@ YUI.add('inspector-relations', function() {
+      Generate the buttons for the status.
+
+      @method _generateButtons
+      @returns {Array} The list of buttons
```
`@returns {Object} The button row component.`

huwshimi commented:
Looks good @anthonydillon :+1: with one QA issue and a few other minor comments. Thanks!

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5167/
Test PASSed.

anthonydillon commented:
@huwshimi thanks for the review. I have fixed all the points you mentioned. Although I believe the QA issue is expected behaviour. When you remove the last committed relation from an application the relation itself goes into uncommitted state and removed from the db. Therefore there is no active relations on the application and without any selected the Remove button is disabled.

jujugui commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/5176/
Test PASSed.

anthonydillon commented:
Thanks :shipit:

jujugui commented:
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
